reco_utils/dataset/pandas_df_utils.py

```diff
@@ -166,3 +167,97 @@ def _convert(field, feature, field_index, field_feature_index_dict):
     return df_new
 
 
+def negative_feedback_sampler(
+    df, 
+    col_user=DEFAULT_USER_COL,
+    col_item=DEFAULT_ITEM_COL,
+    col_label=DEFAULT_LABEL_COL,
+    number_neg_per_user=1,
+    seed=42
+):
+    """Utility function to sample negative feedback from user-item interaction dataset.
+
+    This negative sampling function will take the user-item interaction data to create 
+    binarized feedback, i.e., 1 and 0 indicate positive and negative feedback, 
+    respectively. 
+
+    Negative sampling is used in the literature frequently to generate negative samples 
+    from a user-item interaction data.
+    See for example the neural collaborative filtering paper 
+    https://www.comp.nus.edu.sg/~xiangnan/papers/ncf.pdf
+    
+    Examples:
+        >>> import pandas as pd
+        >>> df = pd.DataFrame({
+            'userID': [1, 2, 3],
+            'itemID': [1, 2, 3],
+            'rating': [5, 5, 5]
+        })
+        >>> df_neg_sampled = negative_feedback_sampler(
+            df, col_user='userID', col_item='itemID', number_neg_per_user=1
+        )
+        >>> df_neg_sampled
+        userID  itemID  feedback
+        1   1   1
+        1   2   0
+        2   2   1
+        2   1   0
+        3   3   1
+        3   1   0
+
+    Args:
+        df (pandas.DataFrame): input data that contains user-item tuples.
+        col_user (str): user id column name.
+        col_item (str): item id column name.
+        col_label (str): label column name. It is used for the generated columns where labels
+        of positive and negative feedback, i.e., 1 and 0, respectively, in the output dataframe.
+        number_neg_per_user (int): number of negative feedback sampled for each user. 
+        seed (int): seed for the random state of the sampling function.
+
+    Returns:
+        pandas.DataFrame: data with negative feedback 
+    """
+    columns = df.columns
+    if col_user not in columns:
+        raise ValueError("Column {} is not in the dataframe".format(col_user))
+
+    if col_item not in columns:
+        raise ValueError("Column {} is not in the dataframe".format(col_item))
+
+    # Get all of the users and items.
+    users = df[col_user].unique()
+    items = df[col_item].unique()
+
+    # Create a dataframe for all user-item pairs
+    user_item_tuples = [(row[col_user], row[col_item]) for _, row in df.iterrows()]
+
+    # Generate feedback for both positive and negative
+    user_item_feedback = [
+        (user, item, 1 if (user, item) in user_item_tuples else 0) for user in users for item in items
+    ]
+    df_all = pd.DataFrame(user_item_feedback, columns=[col_user, col_item, col_label])
+
+    # Take all positive feedback
+    df_pos = df_all[df_all[col_label] == 1]
+
+    # Sample # negative feedback for each user
+    # If the total negative feedback for a user is less than the sampling size, all of the 
+    # negative feedback will be generated.
+    df_neg = df_all[df_all[col_label] == 0]
```

Comment:
Is this different from "remove-seen" or can you regard this as that? If this is something similar, you may also reuse `filter_by()` or use `user_item_pairs(user_item_filter_df=your_positive_samples)` which will return all the user-item pairs except your positive samples.

Reply:
👍